Replace the terminology for shared non-conforming slave/master

YohannDudouit opened the pull request:
This PR improves further the new `Mesh::FaceInformation` by associating `location` and `conformity` to both `elem1` and `elem2` for a given face, this should make the `Mesh::FaceInformation` even easier to process by a human.

This PR also proposes to replace the concepts of `slave` and `master` faces for two main reasons:
- These terms are politically loaded, especially in the US due to its history,
- This terminology is hard to grasp, indeed `slave` and `master` refer to **how** the faces are created in the code, therefore understanding **how** the code works is a prerequisite to understand the documentation (?!). Whereas replacing these concepts by `fine` and `coarse` describes **what** the faces are from a mathematical/geometrical perspective. Furthermore, `slave` and `master` are not sufficient to describe all the possible face configurations, and thus a third concept of `ghost` faces (also not trivial to grasp) is needed to resolve the ambiguity. However, when describing faces in term of `coarse` and `fine`, all the possible face configurations can be described (non-ambiguously) without requiring a third concept.

v-dobrev commented on mesh/mesh.hpp:
```diff
@@ -1272,46 +1275,47 @@ class Mesh
    {
-      FaceLocation location; // Local, shared, boundary
-      FaceConformity conformity; // Conforming, non-conforming master or slave
+      FaceLocation elem_1_location, elem_2_location; // Local, shared, boundary
```
The name `FaceLocation` is not good here -- these are properties of the two adjacent elements, so it is better to call the enum `ElementLocation`.

Also, "shared" location for the element is not a good term to use here -- in MFEM a geometric entity is called "shared" when it belongs to 2 or more ranks. The term we use to describe elements from other ranks that have a common face with an element on this rank is "face-neighbor".

Also, "boundary" location is unnecessary -- for boundary faces we can simply set `elem_1_location = Local` and `elem_2_location = NA`.

v-dobrev replied:
Actually, `elem_1_location` is always going to be `Local`, and only `elem_2_location` can be `Local`, `FaceNeighbor`, or `NA`. So we can just store `elem_2_location`.

YohannDudouit replied:
I know, but it's the same with the orientation, a compiler is smart enough to optimize this, we don't need to do it ourselves.

YohannDudouit replied:
Besides, I don't think it is the same to assume that users will know the convention that `elem_1_location` is always `Local`, and to put this information in the code so that user don't have to know this convention.

pazner replied:
If `elem_1_location` is always guaranteed to be `Local`, you could do something like
```c++
struct FaceInformation
{
   static constexpr FaceLocation elem_1_location = FaceLocation::Local; // Element 1 is always guaranteed to be "Local"
   FaceLocation elem_2_location;
   […]
};
```
That way the invariant is stated explicitly rather than implicitly, and users could still access `elem_1_location` without knowing this detail and it should work correctly.

I think Yohann is right that this should not incur any real runtime or memory cost.

tzanio replied:
I like @pazner's suggestion.

YohannDudouit replied:
I was about to do this change when I realized that it doesn't work with the current enum type, since `elem_1_location` can also be `Boundary`.

A solution that may make everyone happy may be to have:
```c++
enum class ElementLocation { Local, Shared, NA };
```
As @v-dobrev was mentioning we may be able to find a better name than `Shared`, as this describes more the face than a neighboring element. But `Neighbor` will not be very clear I think. `Ghost` could be confusing as "ghost faces" refers only to specific shared non-conforming faces (not all shared non-conforming faces are ghost faces).

And replace the two `FaceConformity` with a unique `FaceConfig` for the face:
```c++
enum class FaceConfig
{
   Boundary,
   Conforming,
   NonConformingFineCoarse, // elem1 is fine, elem2 is coarse
   NonConformingCoarseFine, // elem1 is coarse, elem2 is fine
   NA
};
```

I would put in `FaceConfig::NA` the "ghost master faces", and the "uninitialized ghost faces" because they are undistinguishable from each other and only contain `-1`s (unset values).
And I would also put the "local non-conforming master" faces in `FaceConfig::NA` since only "elem_1_index" is set (everything else is unset), making them mostly useless and only confusing I think. This would prevent users from having to systematically filter them out (and to have to understand that they don't contain usable information), for instance the code in #2195 would become more readable, as I could remove all the recurring:
```c++
      if ( face.IsLocalNonConformingCoarseFine() )
      {
         // We skip local non-conforming coarse-fine faces as they are treated
         // by the corresponding local non-conforming fine-coarse faces.
         continue;
      }
```
MFEM already systematically filters-out "local non-conforming master" faces.

v-dobrev commented on mesh/mesh.hpp:
```diff
@@ -1272,46 +1275,47 @@ class Mesh
+      FaceLocation elem_1_location, elem_2_location; // Local, shared, boundary
+      // elem1 and elem2 face conformity, i.e., conforming, non-conforming
+      // coarse, or non-conforming fine (or NA)
+      FaceConformity elem_1_conformity, elem_2_conformity;
```
These two properties are complementary to each other -- if you know 1 then you know the other.

I suggest we combine them into 1 property:
```c++
   enum class CoarseSide { Conforming, Side1, Side2 };
   CoarseSide coarse_side;
```

Even better: to be more concise and prevent illegal configurations like `elem_2_location = NA` and `coarse_side = Side2`, we can combine `elem_2_location` and `coarse_side` into 1 property:
```c++
   enum class FaceConfig
   {
      Boundary, // always conforming
      LocalConforming, // both sides are local
      LocalNonconforming, // side 1 is the coarse side
      SharedConforming, // side 1 is always the local side
      SharedFineCoarse, // (fine face) side 1 is fine, side 2 is coarse
      SharedCoarseFine, // (fine face) side 1 is coarse, side 2 is fine
      SharedOther // other shared faces, e.g. coarse nonconforming or ghost that is not used
   }
```

v-dobrev replied:
Actually, we need two constants for local non-conforming faces: `LocalFine` and `LocalCoarse`:
```c++
   enum class FaceConfig
   {
      Boundary, // always conforming
      LocalConforming, // both sides are local
      LocalFine, // local fine face, side 1 is coarse
      LocalCoarse, // local coarse face, side 1 is coarse
      SharedConforming, // side 1 is always the local side
      SharedFineCoarse, // (fine face) side 1 is fine, side 2 is coarse
      SharedCoarseFine, // (fine face) side 1 is coarse, side 2 is fine
      SharedOther // other shared faces, e.g. coarse nonconforming or ghost that is not used
   };
```

YohannDudouit replied:
I don't think this is making things more clear, we may want to ask the team's opinion.

v-dobrev replied:
I think seeing, in one place, all possible valid configurations makes things clearer.

YohannDudouit replied:
What you are proposing is missing the point of `GetFaceInformation`, the purpose of `GetFaceInformation` is to have the information available to a program and NOT in the documentation. This is just bad from a design point of view to put information in the documentation when it can be in the code.

v-dobrev replied:
The information provided by `elem_1_location`, `elem_2_location`, and others, can be provided as simple methods, based on the proposed `FaceConfig`.

pazner replied:
I think the idea of enumerating exactly the allowable configurations, and then providing convenience accessors based on that information is a good one. It will prevent illegal configurations (enforced by design), and be more explicit.

YohannDudouit replied:
I do not agree with this, even if I'm sensitive to your argument about illegal configurations, I think this proposition reproduces largely the flaws of `Mesh::FaceInfo`:
- The proposed naming reflects the internals, implementation details into a high-level interface meant to be as human readable as possible. The purpose of `GetFaceInformation` is to abstract the face data structure and implementation by proposing a mathematical/geometrical terminology represented by enum types. This proposition pushes users to understand the data structure and implementation, which is exactly what I'm trying to hide.
- The names in the enumerated type are not transparent as they conflate orthogonal semantics by creating unnecessary neologisms, e.g., `SharedOther` has absolutely no intuitive interpretation. I am completely open to find more appropriate terminology to describe face properties inspired by the scientific literature, but we should not conflate these properties into a single enum. Besides, this approach in general result in a combinatorial explosion of semantics which make it harder for users to understand a face representation.
- For similar reasons as what @samuelpmishLLNL explained [here](https://github.com/mfem/mfem/pull/2195#discussion_r724387693) using different enums makes the code easier to read and to interact with in a debugger. Besides, I could see benefits in modifying the `Mesh::FaceInformation` while debugging which is easier if the different semantics are separated in different enums.

pazner replied:
Agreed that `SharedOther` is opaque. I don't know what it means — are the ghost faces documented?

v-dobrev replied:
The short answer, for users, about `SharedOther` is that these are faces you want to skip -- the other fields in the struct are undefined.

pazner replied:
Is there a longer answer for developers? 🙂 I'm asking because I don't understand this part of the code, but I think it would be valuable to know.

v-dobrev replied:
The answer is that currently these include all ghost faces (as defined by the `ParNCMesh`) that are not shared nonconforming fine faces, see the code around this line: https://github.com/mfem/mfem/blob/bba2c08025c0700ead5965d27d7ed0f8564a0018/mesh/pncmesh.cpp#L1065 However, in `ParMesh`, or in this class, there is no useful information about them. They are simply an artifact of how the `Mesh::faces_info` is constructed right now. In principle, these unused face indices can be filtered out but that is probably not super important.

tzanio replied:
I agree with Yohann that simplicity and human readability are important, but I also like the option to have enumerated possibilities for developers that want to dive in at that level.

Is it possible to keep `elem_1_conformity`, `elem_2_conformity`, etc. as primary and add a method `FaceType` or something that returns either `invalid` or the appropriate option from `FaceConfig`?

v-dobrev commented on mesh/mesh.hpp:
```diff
@@ -1330,38 +1334,46 @@ class Mesh
+                elem_2_conformity==Mesh::FaceConformity::Conforming;
+      }
+
+      /// @brief Return true if the face is a non-conforming fine face.
+      bool IsNonConformingFine() const
```
This is not a good name for this method. If I understand this correctly, this returns `true` if and only if the face is fine nonconforming face AND side 1 is the fine element (which can only happen if the face is shared).

YohannDudouit replied:
I think this is the exact definition in the documentation above:
```
By extension, we name a face a non-conforming coarse or fine face,
if the face belonging to elem1 is a non-conforming coarse of fine face
respectively.
```

v-dobrev replied:
The terms "non-conforming fine face" (aka "non-conforming slave face") and "non-conforming coarse face" (aka "non-conforming master face") are already well defined and used in other places in MFEM. It is better to not introduce a different meaning here.

YohannDudouit replied:
The whole point, CF PR description, is that "non-conforming fine face" is not the same as "non-conforming slave face", and likewise for "non-conforming coarse face" and "non-conforming master face".
Besides, master and slave are not documented as far as I know!

v-dobrev replied:
Assigning different meanings to "non-conforming fine face" and "non-conforming slave face" will be quite confusing I think; similarly with "non-conforming coarse face" and "non-conforming master face".

v-dobrev replied:
One possible name for the method that does not create ambiguity is `HasFineSide1()`.

v-dobrev commented:
@pazner, @tzanio, @psocratis, @dylan-copeland,

If you can take a look and comment on the following discussions, that will be great:
- https://github.com/mfem/mfem/pull/2769#discussion_r788245257
- https://github.com/mfem/mfem/pull/2769#discussion_r788258476
- and also the naming of the methods in class `FaceInformation`.

Thanks!

pazner commented on mesh/mesh.hpp:
```diff
@@ -1259,11 +1259,16 @@ class Mesh
-       a non-conforming slave face, or a non-conforming master face. */
+   /** This enumerated type is used to describe if an element face is a
+       conforming or a non-conforming face. In the case of a non-conforming
+       face, the element can either be the coarse element or the fine element of
+       the non-conforming face.
+       By extension, we name a face a non-conforming coarse or fine face,
+       if the face belonging to elem1 is a non-conforming coarse of fine face
+       respectively. */
    enum class FaceConformity { Conforming,
-                               NonConformingMaster,
-                               NonConformingSlave,
+                               NonConformingCoarse,
+                               NonConformingFine,
                                NA
                              };
```
We had discussed this separately, but I think it might be a bit confusing to call a face a "fine" face if element 1 is the fine element, and "coarse" if element 1 is the coarse element — the way I think about this, the face is equally fine no matter whether element 1 or element 2 is the fine version, it just depends on the numbering of the elements (it is not an inherent property of the face).

Perhaps clearer would be something like
```
NonConformingFineCoarse
NonConformingCoarseFine
```
which would make the symmetry of the situation clearer.

YohannDudouit replied:
Coarse and fine refers to the elements, not the face, as stated in the documentation, and by the variable names called `elem_*`

pazner replied:
I was referring to this in the comment:

> By extension, we name a face a non-conforming coarse or fine face, if the face belonging to elem1 is a non-conforming coarse of fine face respectively.

I think this terminology is confusing.

YohannDudouit replied:
This is actually outdated documentation, it is not used any more, I will remove it.

pazner reviewed (commented):
The improved documentation and naming in this PR is a big improvement. I think "coarse" and "fine" are better and more self-explanatory names than "master" and "slave".

I think what is proposed in this PR is pretty close, but with a little bit more refinement (no pun intended…) we can perfect it.

I think we should try to pick names and terminology that is as self-explanatory as possible, so that users will not be surprised if if they think a term means one thing, but we define it in some other way.

---

I also think the idea of enumerating exactly what the possible/allowable configurations are will be helpful. This is something that is currently implicit in the code, and making it explicit will clear things up.

dylan-copeland commented:
I think this PR has good improvements. One question I have is whether it is possible to replace `elem1` and `elem2` with more descriptive names, because the user will wonder how these two elements are assigned and what can be assumed about them. Maybe this is not possible, if the way these are assigned is too complicated to give simple descriptions (e.g. replace `elem1` with `owningElem`, but that may not be true generally).

YohannDudouit commented:
I replaced the `Elem1` in the function names, which was indeed not very clear nor elegant, and replaced it by the suggested `CoarseFine` and `FineCoarse` names.
